File: Swift/word-pattern.swift
class Solution {
    func wordPattern(_ pattern: String, _ str: String) -> Bool {
        let chars = [Character](pattern)
        let words = str.split { $0 == " "}.map(String.init)

        guard chars.count == words.count else {
            return false
        }
        var dict = [Character: String]()

        for (index, char) in chars.enumerated() {
            let currentWord = words[index]
            if let word = dict[char] {
                if word != currentWord {
                    return false
                }
            } else {
                if dict.values.contains(currentWord) {
                    return false
                } else {
                    dict[char] = currentWord
                }
            }
        }

        return true
    }
}

File: Python/word-pattern.py
class Solution:
    def wordPattern(self, pattern, str):
        """
        :type pattern: str
        :type str: str
        :rtype: bool
        """
        chars = list(pattern)
        words = str.split(' ')

        if len(chars) != len(words):
            return False

        d = dict()

        for index, char in enumerate(chars):
            currentWord = words[index]
            if char in d:
                if d[char] != currentWord:
                    return False
            else:
                if currentWord in d.values():
                    return False
                else:
                    d[char] = currentWord

        return True
